File: Sort/Quick Sort/QuickSort.java
import java.util.Scanner;

public class QuickSort{
   private static int[] array;
   public static void main(String[] args){
      Scanner s = new Scanner(System.in);
      System.out.print("Enter number of elements: ");
      int length = s.nextInt();
      array = new int[length];
      for(int i = 0; i < array.length; i++){
         System.out.print("Enter element n°" + (i+1) + " : ");
         array[i] = s.nextInt();
      }
      
      System.out.println("\nSorting...");
      quickSort(0, length - 1);
      System.out.println("Sort Completed!\n");
      for(int i = 0; i < length; i++)
         System.out.println("Item at index " + i + " is: " + array[i]);
   }
   
   private static void quickSort(int low, int high){
      if(low < high){
         int p = partition(low, high);
         quickSort(low, p);
         quickSort(p + 1, high);
      }
   }
  
  private static int partition(int low, int high){
     int pivot = array[low];
     int i = high + 1;
     for(int j = high; j > low; j--){
        if(array[j] >= pivot){
           i--;
           swap(i, j);
        }
     }
     swap(i-1, low);
     return i-1;
  }
  
  private static void swap(int a, int b){
     int temp = array[b];
     array[b] = array[a];
     array[a] = temp;
  }
}

File: Sort/Quick Sort/QuickSort.kt
object QuickSort {
    private lateinit var array: Array<Int>

    fun quickSort(array: Array<Int>): Array<Int> {
        this.array = array
        sort(0, array.size - 1)
        return this.array
    }

    private fun sort(start: Int, end: Int) {
        if (start < end) {
            val pivot = (start..end).random()
            var temp = array[start]
            array[start] = array[pivot]
            array[pivot] = temp
            var left = start + 1
            var right = end
            while (left <= right) {
                temp = array[left]
                array[left] = array[right]
                array[right] = temp

                while (left <= end && array[left] <= array[start]) {
                    left++
                }
                while (right >= start && array[right] > array[start]) {
                    right--
                }
            }
            temp = array[start]
            array[start] = array[left - 1]
            array[left - 1] = temp
            sort(start, left - 2)
            sort(left, end)
        }
    }

    fun quick3(array: Array<Int>): Array<Int> {
        this.array = array
        sort3(0, array.size - 1)
        return this.array
    }

    private fun sort3(s: Int, e: Int) {
        var start = s
        var end = e
        while (start < end) {
            val pivot = (start..end).random()
            var temp = array[start]
            array[start] = array[pivot]
            array[pivot] = temp

            var m1 = start
            var m2 = start

            var l = start + 1

            while (l <= end) {
                if (array[l] < array[start]) {
                    m1++
                    m2++
                    temp = array[l]
                    array[l] = array[m1]
                    array[m1] = temp
                }

                if (array[l] == array[start] && m2 != l) {
                    m2++
                    temp = array[l]
                    array[l] = array[m2]
                    array[m2] = temp
                }

                l++
            }

            temp = array[start]
            array[start] = array[m1]
            array[m1] = temp

            if (m1 - start < end - m2) {
                sort(start, m1 - 1)
                start = m2 + 1
            } else {
                sort(m2 + 1, end)
                end = m1 - 1
            }
        }
    }
}